src/interfaces/wallet.h

```diff
@@ -255,6 +256,9 @@ class Wallet
     // Remove wallet.
     virtual void remove() = 0;
 
+    // Get a signing provider for the wallet.
+    virtual SigningProvider& getSigningProvider() = 0;
```

Comment:
You may want to check with @achow101 that this doesn't conflict with the descriptor wallet changes.

Reply:
This doesn't interfere with the descriptor wallet changes since a `SigningProvider` was never actually needed in the GUI.

Reply:
But there won't be a singular signing provider for the whole wallet anymore, right?

Reply:
FWIW, I think it would be better to add an `interfaces::Wallet::signMessage()` method instead of adding an `interfaces::Wallet::getSigningProvider()` method. Few reasons:

1) It would allow deduplicating more code for signing messages instead of having it written twice in qt and rpcwallet.

2) It would create a little less work for me with #10102, so I don't have to an add IPC interface around `SigningProvider`, or serialize `SigningProvider` with private key information and send it from the wallet process to the qt process.

3) Just in general I think `rpcwallet.cpp` is too monolothic and has too much important code mixed with UniValue boilerplate. So I'd love to see something like `src/wallet/sign.h` / `src/wallet/sign.cpp` files with a `bool SignMessage(Wallet& wallet, message, options...)` function and start organizing things better.

Anyway, I'm fine with this approach, but wanted to suggest an alternative.

Reply:
@ryanofsky That sounds good to me! I am not sure how to access the `interfaces::Wallet` instance from `rpcwallet.cpp` though. It seems like it only has access to `CWallet` (while the QT side only has access to `interfaces::Wallet`).

Is the idea to add two signmessage functions, where the interface one simply calls down into `m_wallet->SignMessage(..)`?

Reply:
> @ryanofsky That sounds good to me! I am not sure how to access the `interfaces::Wallet` instance from `rpcwallet.cpp` though. It seems like it only has access to `CWallet` (while the QT side only has access to `interfaces::Wallet`).

rpcwallet.cpp shouldn't need to access `interfaces::Wallet` here (and in general) because `interfaces::Wallet` doesn't contain any real functionality, it's just a high level wrapper around selected wallet functions that GUI code uses to indirectly control wallets.

So the idea is just to implement and standalone function in wallet code:

```c++
bool SignMessage(CWallet& wallet, message, options...)
```

and to call this function both from `wallet/rpcwallet.cpp` and `interfaces/wallet.cpp` (just adding a one-line `WalletImpl::signMessage()` wrapper method in the interface code).

> Is the idea to add two signmessage functions, where the interface one simply calls down into `m_wallet->SignMessage(..)`?

Sorry! Just realized I had a typo in https://github.com/bitcoin/bitcoin/pull/16440#discussion_r308805714, which is fixed now but was probably pretty confusing. I was trying to suggest not having a `CWallet::SignMesage(...)` method, but instead defining a standalone `SignMessage(CWallet&, ...)` function. Really either a method or a standalone function would be fine, but the function seemed better because we're gradually breaking CWallet up and pulling functionality out of it, so adding new functionality there without a reason seemed to be moving in wrong direction.

Reply:
@ryanofsky I ended up adding a couple of helper functions to `proof.h` and adding a `proof.cpp` file. The P2PKH-vs-others logic is now all contained inside the proof code, and is called from `rpcwallet`/`misc`/`interfaces/wallet`:

https://github.com/bitcoin/bitcoin/blob/a7632f92a4983d73a9ac460f6e27fadb096492bf/src/script/proof.h#L241-L257

Reply:
As a side result of this, `strMessageMagic` was moved out of validation.h/cpp into proof.cpp, and is no longer public, which is nice.

Reply:
Looks good. The new SignMessageWithSigningProvider makes for fewer wallet changes than what I suggested, which is great!